text/3696-unsized-params-in-traits.md

```diff
@@ -0,0 +1,168 @@
+
+here is some rust-pseudo code:
+```rust
+if !req_function.
+    params
+    .filter(|param| !param.is_sized())
+    .is_empty() {
+        match lint_level {
+            Allow => (),
+            Warn => warn("..."),
+            Deny | Forbid => error("...")
+        }
+    }
+```
+replace the `...` with: 
+```
+The size value for {param_type} is not know at compile time
+# ...
+This was previously accepted by the compiler, but it will become a hard error in the future!
+# if it was the default value of 'deny' the next line would be added
+`#[deny(unsized_param_in_trait)]` on by default
+```
+Obviously the above code isnt actually correct as it doesnt check _which_ param is unsized, it just checks if there is, (you can probably loop over the 'filter' object, and make an individual error for each one)
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- This could cause breaking changes, however the lint gives time for migration
+- This could be intended for `dyn` compatibility (see [rationale-and-alternatives] for a way to fix this)
+This drawback is about receivers, take this example
+```rust
+trait Bar {
+    fn foo(self);
+}
+```
+if `self` was sized checked here, and the value _should_ be consumed, then this code would be impossible without a `Self: Sized` bound, but as you know, adding that bound removes `dyn-Compatibility`
+```rust
+let x: &dyn Bar = &10;
+```
+Produces:
+```
+Bar is not dyn compatible
+# ...
+```
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+This design is best as it helps users to migrate their code before it breaks
+
+Other Considered Designs:
+- as mentioned in [drawbacks], this may be intentional with `receivers` for `dyn` Compatibility
+so another design is dividing this lint into two, one for receivers, and one for regular parameters
+- Not making it a hard error, which could work but it may cause users to make their traits unimplementable.
+- A direct hard error, though not recommended for migration purposes (mentioned above)
+- Leaving it be, though again not recommended as mentioned in [motivation]
+
+The impact of not doing this:
+May cause confusion because a parameter is unsized, and the trait cannot be implemented, which is not good.
+
+This may also cause an `ICE` in some way because the parameters are unsized
+
+# Prior art
+[prior-art]: #prior-art
+
+This feature is not a problem in other languages, weather a type is `Sized` or not, it is abstracted away and you can never know
+
+For example: C++ does not really have unsized types (that i know of)
+Another: C# abstracts the idea of a `Sized` value
+
+Higher level languages do not need to run on the machine directly so there is no need to know the size of a value
```

Comment:
These examples are the opposite of prior art. They explain where there's no prior art. Maybe you can find similar requirements in Rust? I think function pointers are similarly definable with arguments that no real function can have

Reply:
The trivial bounds feature has similar situations, too